Add SPL-402 Starter Kit

astrohackerx opened the pull request:
# SPL-402 Starter Kit

A full-stack starter template for building payment-gated APIs using the SPL-402 protocol on Solana. Enable pay-per-request APIs with 500ms payment verification and zero platform fees.

## What is SPL-402?

SPL-402 is an HTTP 402 protocol built for Solana that enables direct wallet-to-wallet payments for API access. Pay only Solana network fees (~$0.00001) with no intermediaries or platform fees.

**Key Benefits:**

- <1s payment latency
- Zero platform fees
- Direct wallet payments
- Simple integration

Learn more: [spl402.org](http://spl402.org) | [GitHub](http://github.com/astrohackerx/spl402) | [npm](https://www.npmjs.com/package/spl402)

## Features

- Express.js backend with SPL-402 middleware (TypeScript)
- React 19 frontend with Solana Wallet Adapter (TypeScript)
- 4 demo endpoints (free, premium, ultra, enterprise tiers)
- Tailwind CSS UI with real-time payment feedback
- Mainnet-ready configuration

## Quick Start

### Prerequisites

- Node.js 22 or higher
- A Solana wallet (Phantom, Solflare, etc.)
- SOL on mainnet for testing payments

### 1. Server Setup

```bash
cd server
npm install
cp .env.example .env
```

Edit the `.env` file with your configuration:

```env
PORT=3001
RECIPIENT_WALLET=your_solana_wallet_address_here
SOLANA_NETWORK=devnet
SOLANA_RPC_URL=https://api.devnet.solana.com
FRONTEND_URL=http://localhost:5173
```

Start the server:

```bash
npm start
```

### 2. Client Setup

In a new terminal:

```bash
cd client
npm install
cp .env.example .env
```

Edit the `.env` file:

```env
VITE_API_URL=http://localhost:3001
```

Start the development server:

```bash
npm run dev
```

### 3. Test the Application

1. Open http://localhost:5173 in your browser
2. Test the free tier (no wallet required)
3. Connect your Solana wallet
4. Try the premium tiers (requires SOL payment)

## Project Structure

```
spl402-starter-kit/
├── server/          # Express.js backend (TypeScript)
│   ├── index.ts     # SPL-402 middleware & endpoints
│   └── .env.example # Server config
└── client/          # React 19 frontend (TypeScript)
    ├── src/
    │   ├── App.tsx        # Main app logic
    │   ├── main.tsx       # Wallet providers
    │   ├── types.ts       # TypeScript types
    │   └── components/    # Reusable components
    └── .env.example       # Client config
```

## API Endpoints

| Endpoint               | Price     |
| ---------------------- | --------- |
| `/api/free-data`       | Free      |
| `/api/premium-data`    | 0.001 SOL |
| `/api/ultra-premium`   | 0.005 SOL |
| `/api/enterprise-data` | 0.01 SOL  |

## Customization

**Add new endpoints** in `server/index.ts`:

```typescript
// Add route to spl402 config
routes: [{ path: '/api/your-endpoint', price: 0.01 }]

// Add handler
app.get('/api/your-endpoint', (req: Request, res: Response) => {
  res.json({ data: 'Your premium data' })
})
```

**Update UI**: Edit `client/src/App.tsx` or create new components in `client/src/components/`.

## Production Deployment

**Backend** (Railway, Render, DigitalOcean):

- Requires HTTPS and Node.js 18+
- Set environment variables from `.env.example`

**Frontend** (Vercel, Netlify, Cloudflare Pages):

- Update `VITE_API_URL` to production API domain
- Update server `FRONTEND_URL` to match deployed frontend

**Security**: Never commit `.env` files. Use HTTPS in production. Set appropriate CORS origins.

## Mainnet deployment

Devnet by default, to switch to mainnet:

**Server** - Update `.env`:

```env
SOLANA_NETWORK=mainnet-beta
SOLANA_RPC_URL=solana_mainnet_RPC_here
```

**Client** - Update `.env`:

```env
VITE_SOLANA_NETWORK=mainnet-beta
VITE_SOLANA_RPC_URL=solana_mainnet_RPC_here
```

**Important**: Ensure both wallet AND server are on the same network (both devnet or both mainnet)

## Resources

[Documentation](https://spl402.org/docs) | [Solana Docs](https://docs.solana.com) | [Wallet Adapter](https://github.com/solana-labs/wallet-adapter)

## License

MIT

GuiBibeau commented on community/spl402-starter-kit/client/package.json:
```diff
@@ -0,0 +1,29 @@
+    "autoprefixer": "^10.4.18",
+    "postcss": "^8.4.35",
+    "tailwindcss": "^3.4.1",
+    "vite-plugin-node-polyfills": "^0.24.0"
```
are the polyfills needed to support web3.js?

astrohackerx replied:
The client uses vite-plugin-node-polyfills to provide Node.js compatibility layers (Buffer, crypto, stream) required by @solana/web3.js in browser environments

astrohackerx commented:
Hey @GuiBibeau  Thanks for the feedback! Here's what's been updated:

✅ **TypeScript**: Full migration from JavaScript to TypeScript
  - Server: `index.js` → `index.ts` with proper types
  - Client: All `.jsx` → `.tsx` files (App, main, components)
  - Added `types.ts` for shared interfaces

✅ **README**: Simplified

✅ **React 19**: Upgraded from React 18 

✅ **Watch Mode**: Dev script now uses `tsx --watch` instead of duplicating start command

✅ **spl402 v2.0.2**: Updated to latest

✅ **Component Structure**: Refactored monolithic App into modular components


The spl402 SDK requires @solana/web3.js for transaction verification

The client uses vite-plugin-node-polyfills to provide Node.js compatibility layers (Buffer, crypto, stream) required by @solana/web3.js in browser environments

Server Dependencies: The Express.js server maintains @solana/web3.js as a direct dependency for spl402 middleware functionality and cannot migrate to alternative Solana libraries while maintaining protocol compatibility

Ready for review!

astrohackerx commented on community/spl402-starter-kit/package.json:
```diff
@@ -0,0 +1,25 @@
+  "displayName": "SPL-402 Starter Kit",
+  "version": "1.0.0",
+  "description": "Full-stack starter kit for building payment-gated APIs using SPL-402 protocol on Solana",
+  "usecase": "Starter",
```
Done

astrohackerx commented on community/spl402-starter-kit/server/index.js:
```diff
@@ -0,0 +1,130 @@
+import express from 'express'
```
Yep, migrated to ts

astrohackerx commented on community/spl402-starter-kit/client/src/main.tsx:
```diff
@@ -0,0 +1,32 @@
+import { WalletAdapterNetwork } from '@solana/wallet-adapter-base'
+import { WalletModalProvider } from '@solana/wallet-adapter-react-ui'
+import { PhantomWalletAdapter, SolflareWalletAdapter } from '@solana/wallet-adapter-wallets'
+import { clusterApiUrl } from '@solana/web3.js'
```
SPL402Client from spl402 sdk uses @solana/web3.js directly for creating payment transactions

astrohackerx commented:
Hey @GuiBibeau!

I've expanded the capabilities of the SPL402 starter kit by integrating it with OpenRouter. This gives builders simple access to 500+ LLM models with instant pay-per-request payments powered by SPL402. I published it as a separate template: https://github.com/astrohackerx/spl402-OpenRouter

Do you think I should enhance current template with full OpenRouter integration and a UI for accessing LLMs, or keep it minimal so it focuses strictly on demonstrating how SPL402 payments work?

GuiBibeau commented on community/spl402-starter-kit/server/index.ts:
```diff
@@ -0,0 +1,130 @@
+app.use(express.json())
+
+const spl402 = createServer({
+  network: 'mainnet-beta',
```
Should be a parameter like the rpcUrl to match the configuration of the user. 

I'm not sure we should default to mainnet. Users testing this out will prefer to work on devnet first.

astrohackerx commented on community/spl402-starter-kit/client/src/types.ts:
```diff
@@ -0,0 +1,22 @@
+}
+
+export interface LoadingState {
+  [key: string]: boolean
```
Yep, fixed
